docs/contribute/testing.adoc

```diff
@@ -0,0 +1,423 @@
+`Makefile.am`.
+
+
+=== Common test library
+When writing tests for OpenSCAP you should use the common test library which is
+located at link:../../tests/test_common.sh.in[tests/test_common.sh.in].
+
+NOTE: The `configure` script will generate the `tests/test_common.sh` file from
+the link:../../tests/test_common.sh.in[tests/test_common.sh.in] adding
+configuration specific settings.
+
+You will need to source the `tests/test_common.sh` in your entry scripts to
+use the functions which it provides. When sourcing the file, you need to specify
+a relative path to it from the directory with your entry script. For example,
+when your entry script will be in the `tests/API/XCCDF/foo` directory:
+[[app-listing]]
+[source,bash]
+----
+#!/usr/bin/env bash
+
+set -o pipefail
+
+. ../../../test_common.sh
+
+test1
+test2
+...
+----
+
+
+==== Global variables exported by test library
+Always use `$OSCAP` variable instead of the plain `oscap` in your tests when
+calling the `oscap` command line tool. This is because tests might be run with
+`CUSTOM_OSCAP` variable.
+
+You can use `$XMLDIFF` in your tests which will call the
+link:../../tests/xmldiff.pl[tests/xmldiff.pl] script.
+
+It is also possible to do XPath queries using `$XPATH` variable, the usage is:
+[[app-listing]]
+[source,bash]
+----
+$XPATH FILE 'QUERY'
+----
+
+
+==== Setting shell attributes for your test scripts
+It is probably good to set `pipefail` option for all your test scripts so you
+accidentally don't lose non-zero exit statuses of commands in a pipeline:
+[[app-listing]]
+[source,bash]
+----
+set -o pipefail
+----
+
+The next option you can consider is `errexit` which exits your script
+immediately if a command exits with a non-zero status. You might want to use
+this option if tests are somehow dependent and it doesn't make sense to continue
+testing after one test fails.
+[[app-listing]]
+[source,bash]
+----
+set -e
+----
+
+
+==== test_init function
+NOTE: This function should only be called from the entry script which is
+responsible for running all the tests from a specific directory.
+
+The function expects that you provide it with the name of a test log file.
+All tests in a directory will report results into this log file:
+[[app-listing]]
+[source,bash]
+----
+test_init test_api_xccdf_foo.log
+----
+
+NOTE: The specified log file will contain output of both `stdout` and `stderr`
+of every executed test in a directory.
+
+
+==== test_run function
+NOTE: This function should only be called from the entry script which is
+responsible for running all the tests from a specific directory.
+
+The function is responsible for executing a test script file or a function and
+logging its result into the log file created by the `test_init` function.
+[[app-listing]]
+[source,bash]
+----
+test_run "DESCRIPTION" $srcdir/TEST_SCRIPT_FILE|TEST_FUNCTION ARG [ARG...]
+----
+The `$srcdir` variable contains the path to the directory with the entry script.
```

Comment:
I think we could describe the $srcdir and $builddir more in detail somewhere and describe which problems they have. From my experience, I often use them wrong which causes that `make distcheck` doesn't work. I would be nice to describe the difference of `make check ` and `make distcheck` as well.

Reply:
@jan-cerny As this is the gnu autotools specific thing it might be better to add a reference to its documentation, for example the following link:
https://www.gnu.org/software/automake/manual/html_node/Checking-the-Distribution.html

If you have some problem which is not clear from the documentation though, you can describe it here and I will try to add it into the testing guide.

Reply:
It is a nice link, but my problem is that we always use it wrong, we do not run make distcheck, then we submit a PR, and then we wonder why Jenkins fails.

Reply:
not always, but frequently